[sw/crypto] Add P-384 ECDH binary entrypoint

wettermo opened the pull request:
This PR

- adds OTBN binary code for P384 ECDH entrypoint. To be able to have keypair generation and shared key computation in one binary projective add in p384_base.s is optimized by removing redundant instructions. This frees enough IMEM to have a single ECDH entrypoint file.
- adds a corresponding test for ECDH shared key generation.
- adds OTBN binary code to check if a given point is a valid P-384 curve point. For ECDH, to check if a given public key is valid, this binary has to be called in advance of generating a shared key. Unfortunately, due to IMEM limitations, this can't be a part of the ECDH entrypoint and therefore has to be handled separately. It is important for later C Ibex driver development to consider this.
- adds a corresponding test for curve point validation.
- adds arithmetic masking to the p384 scalar mult function. For this, conversion from projective to affine coordinates is separated from the internal scalar mult function. Each function that calls the internal scalar multiplication is adapted accordingly. Corresponding tests are also updated.

jadephilipoom commented on sw/otbn/crypto/p384_a2b.s:
```diff
@@ -133,13 +133,6 @@ p384_arithmetic_to_boolean_mod:
  * clobbered flag groups: FG0
  */
 p384_arithmetic_to_boolean:
-  /* Initialize inputs: in case of randomness in upper part of inputs
-     truncate to 385 bits. */
-  bn.rshi   w19, w19, w31 >> 129
-  bn.rshi   w19, w31, w19 >> 127
-  bn.rshi   w12, w12, w31 >> 129
-  bn.rshi   w12, w31, w12 >> 127
-
```
Is there a reason why these bits will be ignored later on, or is it now a precondition of the function that they be 0?

wettermo replied:
I was searching for possibilities to save imem, and I discovered that in the way the calling functions are written there won't be any randomness in the upper part of the input. Because the calling function only receives actual 384-bit values, there won't be randomness in the more significant bits. Therefore, I removed those lines. (Probably something that can be done for p256 too.)

jadephilipoom commented on sw/otbn/crypto/p384_curve_point_valid.s:
```diff
@@ -0,0 +1,204 @@
+  /* Init all-zero register. */
+  bn.xor    w31, w31, w31
+
+  /* Read the mode and tail-call the requested operation. */
+  la        x2, mode
+  lw        x2, 0(x2)
+
+  addi      x3, x0, MODE_SHARED_KEY
+  beq       x2, x3, p384_curve_point_valid
```
Since there's only one option, I think you can drop the mode parameter here.

wettermo replied:
Makes sense!

jadephilipoom commented on sw/otbn/crypto/p384_curve_point_valid.s:
```diff
@@ -0,0 +1,204 @@
+  csrrs     x2, 0x7c0, x0
+  srli      x2, x2, 3
+  andi      x2, x2, 1
+  bne       x2, x0, _pt_valid
+  unimp
+
+  _pt_valid:
```
Maybe move the jump above, or add more `unimp` repetitions, as an FI protection?

wettermo replied:
Good point, but wouldn't it then also make sense to add more `unimp` instructions after the above jumps?

jadephilipoom replied:
You can, but I think these are less critical from a security standpoint as long as (x, y) still checks out as on-curve.

jadephilipoom commented on sw/otbn/crypto/p384_isoncurve.s:
```diff
@@ -0,0 +1,184 @@
+
+.data
+
+/* Right side of Weierstrass equation */
+.globl r
+.balign 32
+r:
+  .zero 64
+
+/* Left side of Weierstrass equation */
+.globl s
+.balign 32
+s:
+  .zero 64
+
+/* Curve point x-coordinate. */
+.globl x
+.weak x
+.balign 32
+x:
+  .zero 64
+
+/* Curve point y-coordinate. */
+.globl y
+.weak y
+.balign 32
+y:
+  .zero 64
+
```
nit: since you only reference the pointers, I think you shouldn't need the data buffers here.

jadephilipoom replied:
Never mind, I see below that some parts of the code rely on fixed offsets for these buffers...we should really update how the P384 code handles memory! But we talked about that on slack already, it makes sense to do a separate pass for that :slightly_smiling_face:

jadephilipoom commented on sw/otbn/crypto/p384_scalar_mult.s:
```diff
@@ -73,17 +80,101 @@ p384_scalar_mult:
+  /* Arithmetic masking:
+   1. Generate a random mask r
+   2. Subtract masks from projective x coordinate
+      (x, y, z) -> ((x - m) mod p,
```
```suggestion
      (x, y, z) -> ((x - r) mod p,
```

jadephilipoom commented on sw/otbn/crypto/p384_scalar_mult.s:
```diff
@@ -73,17 +80,101 @@ p384_scalar_mult:
+  bn.lid    x2++, 0(x4)
+  bn.lid    x2++, 32(x4)
+
+  /* Move previously stored mask r and z^-1 into inpud WDRs
```
```suggestion
  /* Move previously stored mask r and z^-1 into input WDRs
```

jadephilipoom commented on sw/otbn/crypto/p384_curve_point_valid.s:
```diff
@@ -0,0 +1,204 @@
+ * clobbered registers: x2, x3, x20 to x23, w0 to w17
+ * clobbered flag groups: FG0
+ */
+p384_curve_point_valid:
```
Would it make sense to move this to `p384_isoncurve.s` so it doesn't have to be repeated in both the test and the entrypoint?

wettermo replied:
I thought of that too, and don't know why I decided to not do it. :D 
But you're right, should make the code clearer.

jadephilipoom reviewed (approved):
LGTM! A few nitpicks here and there but nothing major. Thanks, Moritz, I know this one was a ton of work! :muscle:

wettermo commented:
Thanks for the review @jadephilipoom !
I addressed all the points you mentioned above, and did another force push.
In case CI completes and you don't have any additional remarks, it the PR should be mergable :) .
